File: Crack/Detecting/Postprocessing/bfs.py
from typing import List, Tuple
from collections import deque
import numpy as np

def get_neighbors(point: Tuple[int, int], skeleton: np.ndarray) -> List[Tuple[int, int]]:
    """
    주어진 좌표의 8방향 이웃 중 스켈레톤에서 값이 1인 좌표들을 반환
    
    Args:
        point (tuple): 현재 좌표 (x, y)
        skeleton (numpy.ndarray): 스켈레톤 이미지
        
    Returns:
        list: 이웃 좌표 리스트
    """
    x, y = point
    neighbors = []
    
    # 8방향 이웃 검사
    for dx in [-1, 0, 1]:
        for dy in [-1, 0, 1]:
            if dx == 0 and dy == 0:
                continue
            
            nx, ny = x + dx, y + dy
            if 0 <= nx < skeleton.shape[1] and 0 <= ny < skeleton.shape[0]:
                if skeleton[ny, nx] == 1:
                    neighbors.append((nx, ny))
    
    return neighbors

def is_branch_point(point: Tuple[int, int], skeleton: np.ndarray) -> bool:
    """
    해당 좌표가 분기점인지 확인 (3개 이상 연결된 지점)
    
    Args:
        point (tuple): 검사할 좌표 (x, y)
        skeleton (numpy.ndarray): 스켈레톤 이미지
        
    Returns:
        bool: 분기점 여부
    """
    neighbors = get_neighbors(point, skeleton)
    return len(neighbors) >= 3

def is_end_point(point: Tuple[int, int], skeleton: np.ndarray) -> bool:
    """
    해당 좌표가 끝점인지 확인 (1개만 연결된 지점)
    
    Args:
        point (tuple): 검사할 좌표 (x, y)
        skeleton (numpy.ndarray): 스켈레톤 이미지
        
    Returns:
        bool: 끝점 여부
    """
    neighbors = get_neighbors(point, skeleton)
    return len(neighbors) == 1

def select_topleft_start_point(end_points: List[Tuple[int, int]]) -> Tuple[int, int]:
    """
    끝점 리스트에서 가장 왼쪽 위(좌상단)에 있는 점을 선택합니다.
    좌표가 작을수록 왼쪽/위에 있다고 판단합니다.
    
    Args:
        end_points (List[Tuple[int, int]]): 끝점 좌표 리스트
        
    Returns:
        Tuple[int, int]: 선택된 좌상단 시작점 (x, y)
    """
    if not end_points:
        return None
    
    # 왼쪽 위에 있는 점을 찾기 위해 (y, x) 순으로 정렬
    # y 값이 같으면 x 값으로 비교 (y가 작을수록 위, x가 작을수록 왼쪽)
    return min(end_points, key=lambda p: (p[1], p[0]))

def extract_all_paths_with_end_points(skeleton: np.ndarray, end_points: List[Tuple[int, int]]) -> List[List[Tuple[int, int]]]:
    """
    스켈레톤 이미지에서 끝점 리스트를 받아 왼쪽 위에 있는 끝점을 시작점으로 사용하여 모든 경로 추출
    
    Args:
        skeleton (numpy.ndarray): 0 또는 1로 표현된 2D 스켈레톤 이미지
        end_points (List[Tuple[int, int]]): 끝점 좌표 리스트
        
    Returns:
        list: 모든 경로의 리스트 (각 경로는 좌표 튜플의 리스트)
    """
    # 끝점 리스트에서 왼쪽 위에 있는 점을 시작점으로 선택
    start = select_topleft_start_point(end_points)
    
    if start is None or skeleton[start[1], start[0]] == 0:
        return []  # 유효한 시작점이 없으면 빈 리스트 반환
    
    # 결과 경로 저장
    all_paths = []
    paths = []
    
    # 방문한 좌표 추적
    visited = set()
    
    # BFS 탐색을 위한 큐: (현재 좌표, 현재까지의 경로)
    queue = deque([(start, [start])])
    
    while queue:
        current, path = queue.popleft()
        
        # 이미 방문한 좌표라면 스킵
        if tuple(current) in visited:
            continue
            
        visited.add(tuple(current))
        
        # 현재 좌표가 끝점이거나 분기점이면 현재까지의 경로 저장
        if current != start and (is_end_point(current, skeleton) or is_branch_point(current, skeleton)):
            all_paths.append(path)
            
            # 분기점인 경우 계속 탐색 (각 분기를 따로 탐색)
            if is_branch_point(current, skeleton):
                # 분기점에서는 현재 좌표를 새로운 시작점으로 BFS 다시 시작
                neighbors = get_neighbors(current, skeleton)
                for neighbor in neighbors:
                    if neighbor not in visited:
                        queue.append((neighbor, [current, neighbor]))
        else:
            # 일반 경로 포인트의 경우 이웃 좌표로 경로 확장
            neighbors = get_neighbors(current, skeleton)
            for neighbor in neighbors:
                if neighbor not in visited:
                    queue.append((neighbor, path + [neighbor]))

    # 미세 균열 필터링
    for path in all_paths:
        if path and len(path) > 10:
            paths.append(path)
    
    return paths

# # 트리 탐색 너비우선탐색(BFS)
# def neighbors(point: Tuple[int, int]) -> List[Tuple[int, int]]:
#     x, y = point
#     return [(x-1, y), (x+1, y), (x, y-1), (x, y+1),  # 상하좌우
#             (x-1, y-1), (x-1, y+1), (x+1, y-1), (x+1, y+1)]  # 대각선 이동

# def bfs(skeleton: np.ndarray, start: Tuple[int, int], end: Tuple[int, int]):
#     """
#     스켈레톤 기반 BFS 탐색으로 시작점과 끝점 사이의 경로를 찾는 함수

#     Args:
#         skeleton (numpy.ndarray): 0 또는 1로 표현된 2D 스켈레톤 이미지
#         start (tuple): 시작 좌표 (x, y)
#         end (tuple): 끝 좌표 (x, y)

#     Returns:
#         list: 시작점과 끝점 사이의 픽셀값이 1인 좌표 리스트, 경로가 없으면 None
#     """
#     if skeleton[start[1], start[0]] == 0 or skeleton[end[1], end[0]] == 0:
#         return None  # 시작점이나 끝점이 1이 아니라면 경로 없음

#     visited = set()
#     queue = deque([(start, [start])])

#     while queue:
#         current, path = queue.popleft()

#         if current in visited:
#             continue
#         visited.add(current)

#         if current == end:
#             return path

#         for neighbor in neighbors(current):
#             x, y = neighbor
#             if 0 <= x < skeleton.shape[1] and 0 <= y < skeleton.shape[0]:
#                 if skeleton[y, x] == 1 and neighbor not in visited:
#                     queue.append((neighbor, path + [neighbor]))
    
#     return None  # 경로를 찾을 수 없는 경우

# def fnConnectLine(skeleton: np.ndarray, branch_point: List[Tuple[int, int]], end_point: List[Tuple[int, int]]) -> List[List[Tuple[int, int]]]:
#     """
#     스켈레톤 이미지에서 브랜치 포인트와 엔드 포인트를 연결하여 경로를 추출.
#     Args:
#         skeleton (numpy.ndarray): 스켈레톤 이미지 (0과 1로 구성).
#         branch_point (List[Tuple[int, int]]): 브랜치 포인트 리스트.
#         end_point (List[Tuple[int, int]]): 엔드 포인트 리스트.

#     Returns:
#         List[List[Tuple[int, int]]]: 연결된 경로 리스트.
#     """
#     print('브랜치 포인트 : ', branch_point)
#     print('엔드 포인트 : ', end_point)

#     if len(end_point) <= 1:
#         return None

#     visited = set()
#     paths = []
#     filtered_paths = []  

#     def bfs(start: Tuple[int, int]) -> List[Tuple[int, int]]:
#         """
#         BFS를 통해 연결된 경로를 탐색.
#         """
#         queue = deque([start])
#         path = []
#         visited.add(start)

#         while queue:
#             current = queue.popleft()
#             path.append(current)

#             for neighbor in neighbors(current):
#                 x, y = neighbor
#                 if (0 <= x < skeleton.shape[1] and 0 <= y < skeleton.shape[0] and
#                     skeleton[y, x] == 1 and neighbor not in visited):
#                     visited.add(neighbor)
#                     queue.append(neighbor)

#         return path
    
#     def bfs2(start: Tuple[int, int], end: Tuple[int, int]):
#         """
#         스켈레톤 기반 BFS 탐색으로 시작점과 끝점 사이의 경로를 찾는 함수

#         Args:
#             skeleton (numpy.ndarray): 0 또는 1로 표현된 2D 스켈레톤 이미지
#             start (tuple): 시작 좌표 (x, y)
#             end (tuple): 끝 좌표 (x, y)

#         Returns:
#             list: 시작점과 끝점 사이의 픽셀값이 1인 좌표 리스트, 경로가 없으면 None
#         """
#         if skeleton[start[1], start[0]] == 0 or skeleton[end[1], end[0]] == 0:
#             return None  # 시작점이나 끝점이 1이 아니라면 경로 없음

#         visited = set()
#         queue = deque([(start, [start])])

#         while queue:
#             current, path = queue.popleft()

#             if current in visited:
#                 continue
#             visited.add(current)

#             if current == end:
#                 return path

#             for neighbor in neighbors(current):
#                 x, y = neighbor
#                 if 0 <= x < skeleton.shape[1] and 0 <= y < skeleton.shape[0]:
#                     if skeleton[y, x] == 1 and neighbor not in visited:
#                         queue.append((neighbor, path + [neighbor]))
        
#         return None  # 경로를 찾을 수 없는 경우
    
#     def bfs3(start: Tuple[int, int], end: Tuple[int, int]):
#         pass

#     # 1. 브랜치 포인트가 없는 경우: 엔드포인트만 처리
#     if len(branch_point) == 0:
#         for point in end_point:
#             path = bfs(point)
#             paths.append(path)

#     # 2. 브랜치가 1개인 경우: 해당 브랜치와 엔드포인트 연결
#     elif len(branch_point) == 1:
#         for point in end_point:
#             path = bfs2(branch_point[0], point)
#             paths.append(path)

#     # 3. 브랜치가 2개 이상인 경우: 브랜치 간 연결 → 이후 엔드포인트와 가까운 브랜치 연결
#     else:
#         for point in branch_point:
#             path = bfs(point)
#             paths.append(path)

#     for path in paths:
#         if path and len(path) > 1:
#             filtered_paths.append(path)
#             print('경로 : ', path)

#     return filtered_paths

File: 01_rust_test/src/bfs.rs
use std::collections::{VecDeque, HashSet};
use image::{ImageBuffer, RgbImage, Rgb, ImageReader, Limits};

use rand::{Rng, SeedableRng};
use rand::rngs::StdRng;

pub type Point = (usize, usize);


/// 8방향 이웃 중 skeleton[y][x] == 1인 좌표를 반환
fn get_neighbors(point: Point, skeleton: &Vec<Vec<u8>>) -> Vec<Point> {
    let (x, y) = point;
    let mut neighbors = Vec::new();
    let dirs = [-1, 0, 1];
    let height = skeleton.len() as i32;
    let width = skeleton[0].len() as i32;
    let xi = x as i32;
    let yi = y as i32;
    for dx in dirs {
        for dy in dirs {
            if dx == 0 && dy == 0 { continue; }
            let nx = xi + dx;
            let ny = yi + dy;
            if nx >= 0 && nx < width && ny >= 0 && ny < height {
                let ux = nx as usize;
                let uy = ny as usize;
                if skeleton[uy][ux] == 1 {
                    neighbors.push((ux, uy));
                }
            }
        }
    }
    neighbors
}

/// 분기점 여부: 이웃이 3개 이상
fn is_branch_point(point: Point, skeleton: &Vec<Vec<u8>>) -> bool {
    get_neighbors(point, skeleton).len() >= 3
}

/// 끝점 여부: 이웃이 1개
fn is_end_point(point: Point, skeleton: &Vec<Vec<u8>>) -> bool {
    get_neighbors(point, skeleton).len() == 1
}

/// 끝점들 중 y좌표-우선, x좌표-두번째 기준으로 가장 왼쪽/위의 점 선택
fn select_topleft_start_point(end_points: &[Point]) -> Option<Point> {
    if end_points.is_empty() { return None; }
    let mut min_point = end_points[0];
    for &p in end_points.iter().skip(1) {
        if p.1 < min_point.1 || (p.1 == min_point.1 && p.0 < min_point.0) {
            min_point = p;
        }
    }
    Some(min_point)
}

//스켈레톤 가져오기

pub fn load_skeleton_from_png(path: &str, threshold: u8) -> image::ImageResult<Vec<Vec<u8>>> {
    let mut reader = ImageReader::open(path)?;

    let mut limits = Limits::default();
    limits.max_alloc = Some(2 * 1024 * 1024 * 1024); // 2GiB
    reader.limits(limits);

    let img = reader.decode()?.to_luma8();
    // 이하 동일...
     let (w, h) = img.dimensions();
     let mut s = Vec::with_capacity(h as usize);
     for y in 0..h {
         let mut row = Vec::with_capacity(w as usize);
         for x in 0..w {
             let v = img.get_pixel(x, y)[0];
             row.push(if v >= threshold { 1 } else { 0 });
         }
         s.push(row);
     }
     Ok(s)
}




/// BFS로 모든 경로 추출 후 길이 10 이하의 경로는 제거
pub fn extract_all_paths_with_end_points(
    skeleton: &Vec<Vec<u8>>,
    end_points: Vec<Point>,
) -> Vec<Vec<Point>> {
    let start = match select_topleft_start_point(&end_points) {
        Some(p) => p,
        None => return Vec::new(),
    };
    // 시작점이 0인 경우 빈 결과
    if skeleton[start.1][start.0] == 0 {
        return Vec::new();
    }

    let mut all_paths: Vec<Vec<Point>> = Vec::new();
    let mut visited: HashSet<Point> = HashSet::new();
    let mut queue: VecDeque<(Point, Vec<Point>)> = VecDeque::new();
    queue.push_back((start, vec![start]));

    while let Some((current, path)) = queue.pop_front() {
        if visited.contains(&current) {
            continue;
        }
        visited.insert(current);

        let endpoint = is_end_point(current, skeleton);
        let branch = is_branch_point(current, skeleton);
        // 시작점이 아닌 끝점 또는 분기점이면 현재 경로 저장
        if current != start && (endpoint || branch) {
            all_paths.push(path.clone());
            
            // ✅ 여기서 몇 번째 균열인지 출력
            println!(
                "[BFS] {}번째 균열 경로 발견 (길이: {}, 끝점: {:?})",
                all_paths.len(),   // 1번째, 2번째, ...
                path.len(),
                current
            );

            if branch {
                // 분기점이면 현재 좌표에서 다시 BFS 시작
                for neighbor in get_neighbors(current, skeleton) {
                    if !visited.contains(&neighbor) {
                        queue.push_back((neighbor, vec![current, neighbor]));
                    }
                }
                continue; // 각 분기는 새 큐에서 처리
            }
        }
        // 그 외에는 이웃으로 경로 확장
        for neighbor in get_neighbors(current, skeleton) {
            if !visited.contains(&neighbor) {
                let mut new_path = path.clone();
                new_path.push(neighbor);
                queue.push_back((neighbor, new_path));
            }
        }
    }

    // 길이 10 이하 경로 제거
    all_paths.into_iter().filter(|p| p.len() > 10).collect()
}

// ========================================================================= 
// 시각화 코드 

pub fn visualize_paths(
    paths: &Vec<Vec<Point>>,
    width: u32,
    height: u32,
    out_path: &str,
) -> image::ImageResult<()> {
    // 배경 검정
    let mut img: RgbImage = ImageBuffer::from_pixel(width, height, Rgb([0, 0, 0]));

    // 재현 가능한 랜덤 색
    let mut rng = StdRng::seed_from_u64(1234);

    for path in paths {
        let color = Rgb([
            rng.gen_range(80..255),
            rng.gen_range(80..255),
            rng.gen_range(80..255),
        ]);

        for &(x, y) in path {
            // 범위 체크
            if x < width as usize && y < height as usize {
                img.put_pixel(x as u32, y as u32, color);
            }
        }
    }

    img.save(out_path)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_line_path() {
        let skeleton = vec![vec![1u8; 15]];
        let end_points = vec![(0usize, 0usize), (14usize, 0usize)];

        let paths = extract_all_paths_with_end_points(&skeleton, end_points);

        let height = skeleton.len() as u32;
        let width = skeleton[0].len() as u32;

        // cracks.png 라는 파일에 저장
        visualize_paths(&paths, width, height, "cracks.png").unwrap();

        assert_eq!(paths.len(), 1);
        assert!(paths[0].len() > 10);
        assert_eq!(paths[0].first().copied(), Some((0, 0)));
        assert_eq!(paths[0].last().copied(), Some((14, 0)));
    }
}

// #[cfg(test)]
// mod tests {
//     use super::*;

//     #[test]
//     fn simple_line_path() {
//         let skeleton = vec![vec![1u8; 15]];
//         let end_points = vec![(0usize, 0usize), (14usize, 0usize)];

//         let paths = extract_all_paths_with_end_points(&skeleton, end_points);

//         assert_eq!(paths.len(), 1);
//         assert!(paths[0].len() > 10);
//         assert_eq!(paths[0].first().copied(), Some((0, 0)));
//         assert_eq!(paths[0].last().copied(), Some((14, 0)));
//     }
// }